README.md
Ginger
======

[![Latest PyPI version](https://img.shields.io/pypi/v/gingertree.svg)](https://pypi.org/project/gingertree)
[![Build Status](https://github.com/LoicGrobol/ginger/workflows/CI/badge.svg)](https://github.com/LoicGrobol/ginger/actions?query=workflow%3ACI)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

Format conversion and graphical representation of [Universal Dependencies](http://universaldependencies.org) trees.

![2d graphical representation](doc/tree.png)

```text
│
│          ┌─────────────┐
│┌────────┐│             │
││        ││             │┌─────────┐    ┌────┐
││    ┌───│┤         ┌───│┤         │┌──┐│    │
↓│    ↓   ↓│         ↓   ↓│         ↓│  ↓│    ↓
ROOT  Je  reconnais  l'  existence  du  kiwi  .
```

## Installation

First install [Pycairo's dependencies](https://pycairo.readthedocs.io/en/latest/getting_started.html).

If you use Windows, [Pycairo won't install
automatically](https://github.com/pygobject/pycairo/issues/19) anyway but you can install it
manually from [the unoficial wheels](https://www.lfd.uci.edu/~gohlke/pythonlibs/#pycairo).

If you can't or don't want to install Pycairo on your system, skip this and omit \[images\] extras
in the following step.

Install ginger from [Pypi](https://pypi.org/manage/project/gingertree)<a id="footnote-0-1-backref"
href="#footnote-0-1">¹</a>

```console
python3 -m pip install gingertree[images]
```

(Either with `--user` or in a virtualenv but you knew that already, didn't you?)

You can also install it directly from the tip (unstable but usually safe) of the master branch whith

```console
python3 -m pip install git+https://github.com/LoicGrobol/ginger/[images]
```

Test if everything works by running `ginger examples/test.conll`.
The output should be the same as the ASCII-art tree above.

---

<sub><a href="#footnote-0-1-backref">↑</a><a id="footnote-0-1">1.</a> It's `gingertree` because
`ginger` is being squatted ☹ if you know how to fix that, please let me know.</sub>

## Usage

```console
ginger [--from <format>] <in-file> [--to] [<out-file>]
```

### Arguments

- `<in-file>`   input file (in CoNLL-U format), `-` for standard input
- `<out-file>`  output file, `-` for standard input (default: `-`)

### Options

- `-f`, `--from <format>` input file format, see below (default: `guess`)
- `-t`, `--to <format>`   output file format, see below (default: `ascii`)
- `-h`, `--help` Get some help

### Examples

- Print to stdout
  
  ```console
  ginger examples/test.conll
  ```

- Assume CoNLL-X for input format
  
  ```console
  ginger -f conllx spam.conllx
  ```

- Output TikZ code
  
  ```console
  ginger examples/test.conll -t tikz
  ```

- Print to a file

  ```console
  ginger examples/test.conll examples/output.asciiart
  ```

- Pipe in and out
  
  ```console
  cat examples/test.conll | ginger - | less
  ```


### Input formats

- `guess` Try to guess the file format, defaults to CoNLL-U

#### CoNLL

- `conllu` [CoNLL-U v2 format](http://universaldependencies.org/format.html)
- `conllx` [CoNLL-X format](https://web.archive.org/web/20160814191537/http://ilk.uvt.nl:80/conll/)
- `conll2009_gold`  [CoNLL-2009 format](http://ufal.mff.cuni.cz/conll2009-st/task-description.html)
  - Takes only the gold columns into account.
  - The P- and -PRED attributes are preserved in the `misc` attribute of the
    intermediate CoNLL-U tree.
- `conll2009_sys`  [CoNLL-2009 format](http://ufal.mff.cuni.cz/conll2009-st/task-description.html)
  - Takes only the predicted columns into account.
  - The gold columns and the -PRED attributes are preserved in the `misc` attribute of the
    intermediate CoNLL-U tree.

#### Software

Formats used by mainstream NLP tools

- `talismane`  Outputs of
  [Talismane](http://redac.univ-tlse2.fr/applications/talismane/talismane_en.html)
- `mate_gold` Input/Output of
  [mate-tools](http://www.ims.uni-stuttgart.de/forschung/ressourcen/werkzeuge/matetools.en.html)
  (actually an alias for `conll2009_gold`)
- `mate_sys` Input/Output of
  [mate-tools](http://www.ims.uni-stuttgart.de/forschung/ressourcen/werkzeuge/matetools.en.html)
  (actually an alias for `conll2009_sys`)

### Output formats

#### Treebanks

The input must be either the path to an existing file or `-` for standard input. The data that
it contains must be in one of the following formats:

- `guess`           Try to guess the file format, defaults to CoNLL-U
- `conllx`          [CoNLL-X format][2]
- `conllu`          [CoNLL-U format][3]
- `conll2009_gold`  [CoNLL-2009 format][4] (Gold columns only)
- `conll2009_sys`   [CoNLL-2009 format][4] (Predicted columns only)
- `talismane`       Outputs of [Talismane][5]
- `mate_gold`       Alias for `conll2009_gold`, used by [mate-tools][1]
- `mate_sys`        Alias for `conll2009_sys`, used by [mate-tools][1]

**Note**: no real effort is made to preserve informations that are not relevant to Universal
Dependencies, so this might be information-destructive, e.g. if converting from CoNLL-2009 to
itself, the P- attributes will be dropped.

#### Text formats

To use these formats, the output destination must be either a file and thus must not be the path to
an existing directory, or `-` for the standard output.

- `ascii`  ASCII-art (using unicode characters, because, yes, we are subversive)
- `tikz-dependency`   LaTeX code using the `tikz-dependency` package
- `tikz`   TikZ code. Use the `positioning`, `calc` and `shapes.multipart` tikz libraries
  - The output is only the `\tikzpicture` part, not a whole compilable document, there is [an
  example](examples/tree.tex) of such a document in `example`.
  - The code is quite verbose since we chose to rely on TikZ' own arithmetic capabilities in order
  to allow easier edition and reuse of the generated code.

### Image formats

To use these formats, the output destination must be either a directory and thus must not be the
path of an existing file, or `-` for the standard output, in which case the byte streams
corresponding to different trees will be separated by NULL bytes.

- `png`
- `svg`
- `pdf`

[1]: http://www.ims.uni-stuttgart.de/forschung/ressourcen/werkzeuge/matetools.en.html
[2]: https://web.archive.org/web/20160814191537/http://ilk.uvt.nl:80/conll/
[3]: http://universaldependencies.org/format.html
[4]: http://ufal.mff.cuni.cz/conll2009-st/task-description.html
[5]: http://redac.univ-tlse2.fr/applications/talismane/talismane_en.html
[6]: http://universaldependencies.org

## Development

Development and releases on [Github](https://github.com/loic-grobol/ginger).

## Further notes

- When importing non-CoNLL-U treebanks, node forms with spaces are considered to be fixed
  expressions, and are treated according to the [UD
  guidelines](http://universaldependencies.org/u/dep/fixed.html) by splitting them in single-token
  nodes linked by `fixed` dependencies This mean that the exceptions mentioned in [the UDv2
  guidelines](http://universaldependencies.org/v2/segmentation.html) are not supported for those
  imports. They are supported when importing from CoNLL-U, though.

## Licence

This licence (the so-called “MIT License”) applies to all the files in this repository.
See also [LICENSE.md](LICENSE.md).

```text
Copyright 2022 Loïc Grobol <loic.grobol@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
associated documentation files (the "Software"), to deal in the Software without restriction,
including without limitation the rights to use, copy, modify, merge, publish, distribute,
sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or
substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT
OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```


CHANGELOG.md
Changelog
=========

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to
[Semantic Versioning](http://semver.org/).

## [Unreleased]

[Unreleased]: https://github.com/LoicGrobol/ginger/compare/v0.15.0...HEAD

## [0.15.0] - 2022-04-01

[0.15.0]: https://github.com/LoicGrobol/ginger/compare/v0.14.1...v0.15.0

### Added

- Support for the `tikz-dependency` output format

### Changed

- Modernized CI and ops infrastructure
- Modernized packaging
- Ditch docopt and switch to click

## [0.14.1] - 2020-06-02

[0.14.1]: https://github.com/LoicGrobol/ginger/compare/v0.14.0...v0.14.1

### Fixed

- Installation on Windows without cairo ([#10](https://github.com/LoicGrobol/ginger/issues/10))

## [0.14.0] - 2019-12-27

[0.14.0]: https://github.com/LoicGrobol/ginger/compare/v0.13.0...v0.14.0

### Changed

- Dependency on pycairo is optional again
- Usage without installation is not supported anymore

## [0.13.0] - 2018-04-14

### Added

- Image format outputs !!!!!!!!!!!!

## [0.12.0] - 2018-04-14

### Fixed

- `libginger.Tree.root` is now the actual root of the tree
- `libginger.UDNode.space_after` is now correctly true if `misc` is empty
- `libginger.Tree.descendance` is now sorted

### Changed

- Use the regular setuptools install mechanisms instead of our previous homebrewn solution
- Move dist files to [`ginger/`](/ginger/)

## [0.11.0] - 2017-12-27

### Added

- Support for CoNLL-U `sent_id` and `text` metadata
- UD Nodes (`libginger.UDNodes` and subtypes) have a new `space_after: bool` property corresponding
  to UD 'SpaceAfter'
- A tree with extended dependencies in [`test.conll`](examples/test.conll)

### Changed

- `ginger` returns proper exit codes
- Support for `identifier` and `misc` fields for all UD Nodes, including `libginger.MultiTokenNode`
- Support for arbitrary iterables for `deps` in `libginger.Node` constructor
- `_` columns in CoNLL files are now translated to `None` attributes instead of a litteral `_`

### Fixed

- `libginger.Tree.raw_token_sequence` is now actually UD-compliant

  - It doesn't include words that are part of a multi-word token anymore. E.g. only _vámonos_ and
    not _vamos_ and _nos_.
  - It doesn't return the root node anymore.

- `libginger.Tree.word_sequence` is now actually UD-compliant, as it does not include the root node anymore
- `libtreebank.conll…` direct str parsing should work now
- CoNLL-U metadata are now properly read

## [0.10.3] - 2017-08-22

### Fixed

- Importing in tests now works as it should
- Actually ignore lines that should be ignored in treebanks

### Changed

- Full UD for the first example of [`/examples/test.conll`](/examples/test.conll)
- Format the changelog according to [Keep a Changelog v1.0.0](http://keepachangelog.com/en/1.0.0)

## [0.10.2] - 2017-08-22 [YANKED]

## [0.10.1] - 2017-05-04

### Fixed

- Version number in `ginger.py`

## [0.10.0] - 2017-05-04

### Fixed

- Improved release message

## [0.9.2] - 2017-05-04

### Added

- An [auto-release script](tools/release.py) that works in a fashion similar to [apm](https://github.com/atom/apm).

### Fixed

- Running vendored versions from other folders should work

## [0.9.1] - 2017-04-28

### Fixed

- (libginger) `Tree.raw_token_sequence` should work correctly
- Allow vendoring a version when another version is globally installed

## [0.9.0] - 2017-04-27

### Added

- Now support multi-word tokens for CoNLL-U files

### Fixed

- Correctly catch format errors on CoNLL identifiers

## [0.8.1] - 2017-04-14

### Fixed

- CoNLL-U export now deals correctly with **all** empty fields

## [0.8.0] - 2017-04-14

### Fixed

- Fixed bad formatting on parsing error
- Remove obsolete tests
- Fixed Talismane parsing

### Changed

- Hardened CoNLL parsing strictness
- Moved `ParsingError` from `libginger` to `libtreebank`

### Added

- Added basic test for CoNLL-U parsing

## [0.7.0] - 2017-04-07

### Fixed

- Format guessing now plays nice with CoNLL-2009
- Issue proper errors on wrong number of columns for CoNLL-2009
- More graceful handling of SIGINT
- CoNLL2009-sys now properly exports to CoNLL-2009 sys...
- `guess` now correctly guess CoNLL-2009-sys
- CoNLL2009-sys parsing correctly deals with PDEPREL and PFEAT values

### Changed

- In `libtreebank` : the parsers now deal with whole treebank files (as iterables over lines) instead of on a per-tree basis

## [0.6.0] - 2017-04-03

### Fixed

- Issue loading `package.json` in `setup.py` due to encoding
- Force UTF-8 encoding on I/O

### Changed

- Split the CoNLL-2009/mate format into a Gold and a System version

  - The format previously known as `conll2009` is now `conll2009_gold`
  - We have a new `conll2009_sys` that uses predicted columns instead of gold ones

## [0.5.1] - 2017-04-03

### Fixed

- Add dependency on`docopt`

## [0.5.0] - 2017-03-31

### Added

- `mate` format alias for CoNLL-2009
- Treebank outputs (see [README](README.md))

### Changed

- Improve [test.conll](/examples/test.conll) to be an actual treebank instead of a single tree.

### Fixed

- `libtreebank`

  - Trying to parse CoNLL-[UX] files with more or less than 10 columns now throw a proper `ParsingError`.

## [0.4.1] - 2017-03-29

### Changed

- Try to conform to [Keep a Changelog](http://keepachangelog.com/).

### Bugfix

- Properly add EOL at the end of outputs
- Fixed issues with treebank files starting with a blank line

## [0.4.0] - 2017-03-17

### New features

- Support the [CoNLL-2009 format](http://ufal.mff.cuni.cz/conll2009-st/task-description.html)

### API change

- Minor change in the representation of `Node`s

### Bugfix

- Fix invocation from installed package

## [0.3.0] - 2017-03-13

### New features

- Support [Talismane](http://redac.univ-tlse2.fr/applications/talismane/talismane_en.html) outputs as inputs

### Bugfix

- CoNLL-U parsing now support comment lines, empty nodes and multi-word tokens

ginger/ginger.py
#! /usr/bin/env python3
r"""Format conversion and graphical representation of [Universal Dependencies][6] trees.

## Usage:
  ginger [--from <format>] <origin> [--to <format>] [<destination>]

## Arguments:
  <origin>       input origin `-` for standard input
  <destination>  output destination, `-` for standard output [default: -]

See below for details on the authorized types of input and output.

## Options:
  -f, --from <format>  input format, see below [default: guess]
  -t, --to <format>    output format, see below [default: ascii]
  -h, --help           Show this screen.

## Input formats:
The input must be either the path to an existing file or `-` for standard input. The data that
it contains must be in one of the following formats:

  - `guess`           Try to guess the file format, defaults to CoNLL-U
  - `conllx`          [CoNLL-X format][2]
  - `conllu`          [CoNLL-U format][3]
  - `conll2009_gold`  [CoNLL-2009 format][4] (Gold columns only)
  - `conll2009_sys`   [CoNLL-2009 format][4] (Predicted columns only)
  - `talismane`       Outputs of [Talismane][5]
  - `mate_gold`       Alias for `conll2009_gold`, used by [mate-tools][1]
  - `mate_sys`        Alias for `conll2009_sys`, used by [mate-tools][1]

## Output formats:

### Text formats
To use these formats, the output destination must be either a file and thus must not be the path to
an existing directory, or `-` for the standard output.

  - `ascii`  ASCII-art (using unicode characters, because, yes, we are subversive)
  - `tikz`   TikZ code. Use the `positioning`, `calc` and `shapes.multipart` tikz libraries

### Image formats
To use these formats, the output destination must be either a directory and thus must not be the
path of an existing file, or `-` for the standard output, in which case the byte streams
corresponding to different trees will be separated by NULL bytes.

  - `png`
  - `svg`
  - `pdf`

## Example:
  `ginger -f conllu input.conll -t tikz output.tex`

[1]: http://www.ims.uni-stuttgart.de/forschung/ressourcen/werkzeuge/matetools.en.html
[2]: https://web.archive.org/web/20160814191537/http://ilk.uvt.nl:80/conll/
[3]: http://universaldependencies.org/format.html
[4]: http://ufal.mff.cuni.cz/conll2009-st/task-description.html
[5]: http://redac.univ-tlse2.fr/applications/talismane/talismane_en.html
[6]: http://universaldependencies.org
"""

__version__ = 'ginger 0.11.0'

import contextlib
import logging
import pathlib

import signal
import sys

import itertools as it
import typing as ty

from docopt import docopt


try:
    from . import libtreebank
    from . import libtreerender
    from . import __version__
except ImportError:
    # Usual frobbing of packages, due to Python's insane importing policy
    if __name__ == "__main__" and __package__ is None:
        package_root = pathlib.Path(__file__).resolve().parents[1]
        sys.path.insert(0, str(package_root))
    from ginger import libtreebank
    from ginger import libtreerender
    from ginger import __version__

signal.signal(signal.SIGPIPE, signal.SIG_DFL)
logging.basicConfig(level=logging.INFO)


def sigint_handler(signal, frame):
    logging.error('Process interrupted by SIGINT')
    sys.exit(0)


signal.signal(signal.SIGINT, sigint_handler)


# Thanks http://stackoverflow.com/a/17603000/760767
@contextlib.contextmanager
def smart_open(filename: str = None, mode: str = 'r', *args, **kwargs):
    if filename == '-':
        if 'r' in mode:
            stream = sys.stdin
        else:
            stream = sys.stdout
        if 'b' in mode:
            fh = stream.buffer
        else:
            fh = stream
    else:
        fh = open(filename, mode, *args, **kwargs)

    try:
        yield fh
    finally:
        try:
            fh.close()
        except AttributeError:
            pass


def directory_multi_output(path: ty.Union[pathlib.Path, str],
                           data: ty.Iterable[ty.Union[str, bytes]],
                           name_format: str = '{i}'):
    '''
    Write the elements of `data` to individual files in `path`.

    The file names will be `n=name_format.format(i=<number>)` where `<number>` is the first integer
    such that `n` is not an existing file name in `path`.

      - `path` **must not** be the path to an existing file.
    '''
    path = pathlib.Path(path)  # Enforce `path`'s type
    if path.exists() and not path.is_dir():
        raise IOError('{path} is an existing non-directory file.'.format(path=path))
    if not path.exists():
        path.mkdir(parents=True)

    # Reuse the same iterator to find available names, to save on calls to `path.iterdir()`
    names = (name_format.format(i=i) for i in it.count())
    for file_content, file_name in zip(data, (n for n in names if not (path/n).exists())):
        try:
            (path/file_name).write_bytes(file_content)
        except TypeError:
            (path/file_name).write_text(file_content)


def stream_multi_output(stream: ty.BinaryIO,
                        data: ty.Iterable[bytes],
                        separator: ty.Union[bytes, str] = b'\x00'):
    '''Write the elements of `data` to `stream`, separated by `separator`.'''
    separator = bytes(separator)
    data_iter = iter(data)

    # Write the first element without separator before it
    stream.write(next(data_iter))
    for file_content in data_iter:
        stream.write(separator)
        stream.write(file_content)


def main_entry_point(argv=None):
    arguments = docopt(__doc__, version=__version__, argv=argv)
    # Since there are no support for default positional arguments in
    # docopt yet. Might be useful for complex default values, too
    if arguments['<destination>'] is None:
        arguments['<destination>'] = '-'

    with smart_open(arguments['<origin>'], encoding='utf8') as in_stream:
        in_lst = list(in_stream.readlines())

    if arguments['--from'] == 'guess' or arguments['--from'] is None:
        arguments['--from'] = libtreebank.guess(in_lst)

    parser, _ = libtreebank.formats.get(arguments['--from'], None)

    if parser is None:
        logging.error(f'{arguments["--to"]!r} is not supported as an input format')
        return 1

    treebank = parser(in_lst)

    # Binary outputs
    if arguments['--to'] in {'png', 'svg', 'pdf'}:
        if arguments['--to'] == 'png':
            out_bytes_lst = [libtreerender.to_png(t) for t in treebank]
        if arguments['--to'] == 'svg':
            out_bytes_lst = [libtreerender.to_svg(t) for t in treebank]
        if arguments['--to'] == 'pdf':
            out_bytes_lst = [libtreerender.to_pdf(t) for t in treebank]
        if arguments['<destination>'] == '-':
            with smart_open(arguments['<destination>'], 'wb') as out_stream:
                stream_multi_output(out_stream, out_bytes_lst)
        else:
            directory_multi_output(arguments['<destination>'], out_bytes_lst,
                                   name_format=f'{{i}}.{arguments["--to"]}')
    # Text outputs
    else:
        # Text-based graphics
        if arguments['--to']== 'tikz':
            out_lst = [libtreerender.tikz(t) for t in treebank]

        elif arguments['--to'] == 'ascii':
            out_lst = [libtreerender.ascii_art(t) for t in treebank]

        # Treebank
        else:
            _, formatter = libtreebank.formats.get(arguments['--to'], None)

            if formatter is None:
                logging.error(f'{arguments["--to"]!r} is not supported as an output format')
                return 1

            out_lst = [formatter(t) for t in treebank]

        with smart_open(arguments['<destination>'], 'w', encoding='utf8') as out_stream:
            for t in out_lst[:-1]:
                out_stream.write(t)
                out_stream.write('\n\n')
            out_stream.write(out_lst[-1])
            out_stream.write('\n')


if __name__ == '__main__':
    sys.exit(main_entry_point())
